alternator/streams.cc

```diff
@@ -180,5 +180,255 @@ future<alternator::executor::request_return_type> alternator::executor::list_str
     return make_ready_future<executor::request_return_type>(make_jsonable(std::move(ret)));
 }
 
+struct shard_id {
+    utils::UUID table;
+    db_clock::time_point time;
+    cdc::stream_id id;
+
+    shard_id(utils::UUID tab, db_clock::time_point t, cdc::stream_id i)
+        : table(tab)
+        , time(t)
+        , id(i)
+    {}
+    shard_id(const sstring&);
+
+    friend std::ostream& operator<<(std::ostream& os, const shard_id& id) {
+        return os << id.table << std::hex
+            << ':' << id.time.time_since_epoch().count()
+            << ':' << id.id.first()
+            << ':' << id.id.second()
+            ;
+    }
+};
+
+shard_id::shard_id(const sstring& s) {
+    auto i = s.find(':');
+    auto j = s.find(':', i + 1);
+    auto k = s.find(':', j + 1);
+
+    if (i == sstring::npos || j == sstring::npos || k == sstring::npos) {
+        throw api_error("Invalid Stream Id", sstring(s));
+    }
+
+    table = utils::UUID(s.substr(0, i));
+    time = db_clock::time_point(db_clock::duration(std::stoull(s.substr(i + 1, j), nullptr, 16)));
+    id = cdc::stream_id(std::stoull(s.substr(j + 1, k), nullptr, 16), std::stoull(s.substr(k + 1), nullptr, 16));
+}
+}
+
+template<typename ValueType>
+struct rapidjson::internal::TypeHelper<ValueType, alternator::shard_id>
+    : public from_string_helper<ValueType, alternator::shard_id>
+{};
+
+namespace alternator {
+
+enum class stream_view_type {
+    KEYS_ONLY,
+    NEW_IMAGE,
+    OLD_IMAGE,
+    NEW_AND_OLD_IMAGES,
+};
+
+std::ostream& operator<<(std::ostream& os, stream_view_type type) {
+    switch (type) {
+        case stream_view_type::KEYS_ONLY: os << "KEYS_ONLY"; break;
+        case stream_view_type::NEW_IMAGE: os << "NEW_IMAGE"; break;
+        case stream_view_type::OLD_IMAGE: os << "OLD_IMAGE"; break;
+        case stream_view_type::NEW_AND_OLD_IMAGES: os << "NEW_AND_OLD_IMAGES"; break;
+        default: throw std::invalid_argument(std::to_string(int(type)));
+    }
+    return os;
+}
+std::istream& operator>>(std::istream& is, stream_view_type& type) {
```

Comment:
json marshalling. See the helpful type helper in the top of the file. It uses lexical cast.

Reply:
Isn't all this input stream thing significantly less efficient than just calling a function which can work on string_views and check if it's "KEYS_ONLY" or "NEW_IMAGE and such - instead of reading strings from input streams, a gazillion allocations, and so on?